Manage cilium-health endpoint from a controller

joestringer opened the pull request:
```release-note
Fix daemon crash on startup when cilium-health endpoint takes a long time to launch
```

The cilium-health endpoint's lifecycle was previously not handled particularly well: The main goroutine in Cilium would rush to get it up, providing it just five seconds to start before failing out and causing the main daemon to exit. Under high load situations, eg there's a lot of endpoints to restore on startup, this could be exceeded for no reason other than lack of CPU resources available to launch the health endpoint.

By splitting this out into a controller, it can be handled more gracefully and unblock the main goroutine to continue to (attempt to) serve the API. Furthermore, we can avoid using `log.Fatal()` with regards to the launching of the cilium-health endpoint but still represent failures to run this endpoint in a user-visible way, in the status output:

```
$ cilium status --all-controllers
KVStore:                Ok         Consul: 172.17.0.3:8300
ContainerRuntime:       Ok         docker daemon: OK
Kubernetes:             Disabled
Cilium:                 Ok         OK
NodeMonitor:            Disabled
Cilium health daemon:   Ok
IPv4 address pool:      3/65535 allocated
IPv6 address pool:      2/65535 allocated
Controller Status:      10/10 healthy
  Name                                 Last success   Last error   Count   Message
  cilium-health-ep                     6s ago         never        0       no error
...
```

Note that to reasonably introduce a controller to handle the lifecycle of the cilium-health endpoint, some amount of refactoring was required. I've prepared this PR with discrete patches for each functional or non-functional change, so it may assist to review this patch-by-patch rather than as a whole PR at once.

Fixes: #4242

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/4498)
<!-- Reviewable:end -->

joestringer committed:
health-ep: Add health EP to manager later

After connectivity is properly set up, add the endpoint to the endpoint
manager. This should reduce regeneration of the endpoint because it will
have a valid identity before the daemon starts managing it.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
health-ep: Depend on option package for config

These options were recently refactored into a singleton package which
the cilium-health launch package already refers to. Rather than passing
the argument around, refer directly to the package version of the
config.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
health-ep: Refactor cleanup logic

Refactor the cleanup of pre-existing cilium-health endpoint to a
separate function. This will simplify upcoming commits.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
health-ep: Refactor error handling to caller

Rather than calling log.Fatal() directly inside the health endpoint
launch code, format an error and return it to the caller, then call
Fatal from the caller. This should have no operational impact.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
health-ep: Rely on pidfile for deferred cleanup

Rather than launching the cilium-health daemon with a context and
passing the cancel function around to callers and so on, instead just
invoke the health endpoint cleanup logic directly, which attempts to
find the health endpoint based on its pidfile, and delete its veth based
on commonly known device name.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
daemon: Controllerize cilium-health endpoint

Add a controller to manage the lifecycle of the cilium-health endpoint.

Fixes: #4242

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
health-ep: Bump timeouts for endpoint readiness

Rather than checking whether the cilium-health endpoint has launched
every 100ms for 5s, now that the health endpoint is managed via a
controller we can be a bit more casual - check once a second for a
minute. If it really doesn't start for a whole minute, then start to
raise the alarm.

Signed-off-by: Joe Stringer <joe@covalent.io>

raybejjani commented on cilium-health/launch/endpoint.go:
```diff
@@ -102,10 +118,59 @@ func configureHealthRouting(netns, dev string, addressing *models.NodeAddressing
+	if pid, err := ioutil.ReadFile(pidfile); err == nil {
+		// Existing cilium-health exists, kill it so we can create a
+		// new one and steal all its logs (om nom nom!)
+		pidInt, err := strconv.Atoi(string(pid))
```
This shouldn't ever error but it's a file on-disk so who knows. It might be worth returning early with an error log when `Atoi` fails (or break out of the if).
Separately, perhaps some of this can be moved to the pidfile package? health uses it already, I think.

raybejjani commented on cilium-health/launch/endpoint.go:
```diff
@@ -102,10 +118,59 @@ func configureHealthRouting(netns, dev string, addressing *models.NodeAddressing
-func LaunchAsEndpoint(owner endpoint.Owner, hostAddressing *models.NodeAddressing, opts *option.BoolOptions) context.CancelFunc {
+func LaunchAsEndpoint(owner endpoint.Owner, hostAddressing *models.NodeAddressing) error {
+
+	CleanupEndpoint(owner)
```
This is an unexpected side-effect. Since we only call `LaunchAsEndpoint` in a few places, we can always just directly call `CleanupEndpoint` there. Otherwise, if you're really worried about one day launching without cleanup could you update the comment, at least.

raybejjani commented on daemon/main.go:
```diff
@@ -718,9 +719,22 @@ func runDaemon() {
+				addressing := d.getNodeAddressing()
+				if err = health.PingEndpoint(); err != nil {
+					return health.LaunchAsEndpoint(d, addressing)
+				}
```
Could you add a comment explaining that we're using `PingEndpoint` as a proxy for whether the endpoint is running?
As alternatives, we can use a persistent boolean defined outside this function (it will be heap allocated automatically, and there is only one instance of this controller ever so no races occur). A more clear scheme would be to use a sync.Once. I'm not sure if you anticipate having to relaunch the healthendpoint more than once. If so, perhaps it's better yet to check that the pid is still running and is our child?

joestringer replied:
Sure.

I figured that sync.Once could be used for this purpose, but it seemed more complicated than necessary here. It's most likely that any error that occurs while trying to contact the health ep is because of something that can be remedied by restarting the EP. Treating the first initialisation as just another such case means we don't have to single out the initialisation case and handle it differently. Also, even if this assumption is wrong, the controller model should isolate and represent the persistent error in the `cilium status` output nicely.

When Cilium is deployed as a systemd service, on upgrade we also want to kill the version of this endpoint that isn't the current child.

raybejjani commented on cilium-health/launch/endpoint.go:
```diff
@@ -102,10 +118,59 @@ func configureHealthRouting(netns, dev string, addressing *models.NodeAddressing
+
+// CleanupEndpoint attempts to kill any existing cilium-health endpoint and
+// clean up its devices and pidfiles.
+func CleanupEndpoint(owner endpoint.Owner) {
```
Since we create a new veth for the new endpoint the worst case scenario here is that we leak older health endpoints, right? They would never exit but also never see any traffic?

joestringer replied:
Right. This patch is relying on the pidfile being created by the health agent, and that we can use that to determine the old instance PID and terminate it.

For the veth, it's always named the exact same so we can destroy and recreate it.

raybejjani replied:
Hmm, so how would this work out if cilium-agent crashes and the health endpoint is orphaned? I guess it's like deleting an open file? the old veth remains around for the oblivious old cilium-health and we all move on?

joestringer replied:
I believe that if you delete one side of a veth, it cleans up the whole veth and nothing retains a reference to it. Sure, if a packet is currently being forwarded across it, it won't be cleaned up until after the packet has been transmitted, but I don't think it can end up being orphaned long-term.

The namespace for the old endpoint on the other hand would be treated like this, but it is bound to the lifetime of the process through the trap implementation in the spawn_netns script. When the process is killed, it's corresponding namespace should be cleaned up.

(Maybe we can simplify this over time, but that would probably be exceeding the scope of this bugfix)

raybejjani commented on cilium-health/launch/endpoint.go:
```diff
@@ -42,6 +44,20 @@ import (
+	healthPidfile = "health-endpoint.pid"
+
+	// client is used to ping the cilium-health endpoint as a health check.
+	client *healthPkg.Client
```
I don't see an easy way to do this better, but this will be slightly racey between callers of `PingEndpoint` and setting this in `LaunchAsEndpoint`. The only other way I see is to make the client a parameter here and have callers worry about creating it at the correct time.

joestringer replied:
That's why the ping endpoint returns an error if the client is nil, it always hits this the very first time with the current code. I thought about adding a struct to manage the lifetime of this client and let the caller use it and so on, that's probably the most idiomatic way to handle this. Maybe I should just go ahead and do that.

raybejjani replied:
Oh, I mean it would trigger the race detector, and arbitrarily fail that. I suppose it's possible to get a partial read of the variable but I'm pretty sure that won't happen.

How important is it to only have one of these? I imagine the overhead of a few is pretty low. In any case a sync.Once to manage the global singleton would work, but there's a really annoying sequencing requirement here, since I'm guessing it will fail to create if the health endpoint is down, and would need to be recreated on errors?

joestringer replied:
Oh, I think it's only ever touched from the controller goroutine right now. But that means it's not a thread-safe API.

We only need one of them right now, the thing I was trying to avoid is creating a new connection for every request. Yes, it will block and time out if the endpoint is down - which is also why it seemed like a good way to choose when to recreate the endpoint.

raybejjani replied:
right. I think a comment would be good enough overall.

I've been solving this style of problem elsewhere by just making the needed stuff just above the `UpdateController` call and using those in the DoFunc/StopFunc. It's a little yucky, and you have to be careful about reassigning things but it does work. In this case it strikes me as a bad thing since this function is so large to begin with (in k8s land I had a `startThisK8sController` functions that did all this in one scope). I'm not sure that it's worth it, overall.

joestringer committed:
pidfile: Refactor kill by pidfile into pidfile

Refactor process killing from the health endpoint launch code into the
pidfile package so that more idiomatic code style can be used for
handling failures consistently.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
health-ep: Document threadsafety

This package should basically just be used from a single goroutine,
specifically the one that is created for the manager of the
cilium-health endpoint. Document the threadsafety concerns.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
daemon: Allow endpoint to be freed without releasing IP

An upcoming commit will delete the cilium-health endpoint in the daemon,
but doesn't want to put its IP back into the pool. Refactor this
slightly and document the behaviour.

Signed-off-by: Joe Stringer <joe@covalent.io>

ianvernon commented on cilium-health/launch/endpoint.go:
```diff
@@ -102,10 +113,48 @@ func configureHealthRouting(netns, dev string, addressing *models.NodeAddressing
+	pidfile := filepath.Join(option.Config.StateDir, healthPidfile)
+	if pid, err := ioutil.ReadFile(pidfile); err == nil {
+		// Existing cilium-health exists, kill it so we can create a
+		// new one and steal all its logs (om nom nom!)
```
Nice log message :)

ianvernon reviewed (commented):
Well done!! LGTM :D

ianvernon reviewed (approved):
"Commented" instead of "approved" - approving for real this time.
